test/integration/deployment/util.go

```diff
@@ -0,0 +1,330 @@
+			reason = fmt.Sprintf("New replica set %q doesn't have DefaultDeploymentUniqueLabelKey", newRS.Name)
+			d.t.Logf(reason)
+			return false, nil
+		}
+		// Check revision of this deployment, and of the new replica set of this deployment
+		if deployment.Annotations == nil || deployment.Annotations[deploymentutil.RevisionAnnotation] != revision {
+			reason = fmt.Sprintf("Deployment %q doesn't have the required revision set", deployment.Name)
+			d.t.Logf(reason)
+			return false, nil
+		}
+		if deployment.Spec.Template.Spec.Containers[0].Image != image {
+			reason = fmt.Sprintf("Deployment %q doesn't have the required image set", deployment.Name)
+			d.t.Logf(reason)
+			return false, nil
+		}
+		if newRS.Annotations == nil || newRS.Annotations[deploymentutil.RevisionAnnotation] != revision {
+			reason = fmt.Sprintf("New replica set %q doesn't have the required revision set", newRS.Name)
+			d.t.Logf(reason)
+			return false, nil
+		}
+		if newRS.Spec.Template.Spec.Containers[0].Image != image {
+			reason = fmt.Sprintf("New replica set %q doesn't have the required image set", newRS.Name)
+			d.t.Logf(reason)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err == wait.ErrWaitTimeout {
+		d.logReplicaSetsOfDeployment(nil, newRS)
+		err = fmt.Errorf(reason)
+	}
+	if newRS == nil {
+		return fmt.Errorf("deployment %q failed to create new replica set", deploymentName)
+	}
+	if err != nil {
+		return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deployment.Spec.Template.Spec.Containers[0].Image, newRS.Name, newRS.Annotations[deploymentutil.RevisionAnnotation], newRS.Spec.Template.Spec.Containers[0].Image, revision, image, err)
+	}
+	return nil
+}
+
+func (d *deploymentTester) markPodsReady() error {
+	ns := d.deployment.Namespace
+	selector, err := metav1.LabelSelectorAsSelector(d.deployment.Spec.Selector)
+	if err != nil {
+		return err
+	}
+	pods, err := d.c.Core().Pods(ns).List(metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		return err
+	}
+	for i := range pods.Items {
+		pod := pods.Items[i]
+		if !podutil.IsPodReady(&pod) {
+			addPodConditionReady(&pod, metav1.Now())
+			if _, err := d.c.Core().Pods(ns).UpdateStatus(&pod); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (d *deploymentTester) waitForDeploymentStatusValid() error {
+	var (
+		oldRSs, allOldRSs, allRSs []*v1beta1.ReplicaSet
+		newRS                     *v1beta1.ReplicaSet
+		deployment                *v1beta1.Deployment
+		reason                    string
+	)
+
+	name := d.deployment.Name
+	err := wait.Poll(pollInterval, pollTimeout, func() (bool, error) {
+		var err error
+		// Manually update Deployment pods status to ready
+		err = d.markPodsReady()
```

Comment:
This seems like it should be outside of this poll loop - you should also make `markPodsReady` to never error out.